keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,447 @@
+| Can reuse existing libraries (client-go, helpers in documentation) | no | yes|
+| Support for streaming | yes | yes|
+| Strongly typed | yes | no |
+| Versions support | yes | yes |
+
+Given the above table, we believe that a gRPC API is the best path forward.
+
+### Can we integrate with PodResource API?
+
+|   | PodResource API | Status API |
+| - | --------------- | ------------- |
+| What information is returned with API | Container resources (devices, cpu_ids, memory, dynamic resources) | for Static and Regular Pods pod readiness Conditions and UID  |
+| Who is the owner of the data returned by API | Kubelet, those data are not reported to Control Plane | static Pods - Kubelet, regular Pods - Kubelet, data are reported to kube-API |
+| What is the alternative for the API | reading the state files resource manager use | for regular Pods - fetching pod's data with kube-API, for static Pods - none |
+| Control Plane dependency | no | no |
+| Authorization | none, using unix socket to limit the clients to local workloads | none, using unix socket to limit the clients to local workloads |
+| Access | Read-Only | Read-Only|
+| Clients type / use cases | Device Plugins, CSI, Topology aware scheduling | CNI, monitoring and observability tools |
+
+The PodResource API includes an entirely unrelated set of information that is
+unlikely to be of use to the set of clients that would benefit from
+understanding Pod readiness. We propose creating a new API for this purpose.
+
+### Risks and Mitigations
+
+This API is read-only, which removes a large class of risks.
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests to the API impacting the Kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. There is no need to authorize the clients to access this API. |
+| Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) | High | This API should serve only complete information about workloads readiness. If Kubelet is in the init phase and not all pod’s readiness information is known, then the API should report the error. |
+| Kube-API is down or unreachable | Low | Kube-API availability should not impact this API. When the control plane is down or unreachable but Kubelet is working properly this API should return most recent data about local Pods readiness that are computed by Kubelet even if those data were not reported or accepted by Kube-API. |
```

Comment:
I'd like to add one more risk associated with exposing all node's pods readiness:
- enabling listing all pods on a given node (through their readiness) can expose information about what workloads are running in the node/cluster. The API rate limiting mentioned above is one mitigation, but very likely other might need to be added too.

Reply:
done